src/lsptoolshost/buildDiagnosticsService.ts

```diff
@@ -0,0 +1,146 @@
+    /** All the build results sent by the DevKit extension. */
+    private _allBuildDiagnostics: Array<[vscode.Uri, vscode.Diagnostic[]]> = [];
+
+    /** The diagnostic results from build displayed by VS Code. When live diagnostics are available for a file, these are errors that only build knows about.
+     * When live diagnostics aren't loaded for a file, then these are all of the diagnostics reported by the build.*/
+    private _diagnosticsReportedByBuild: vscode.DiagnosticCollection;
+
+    constructor(buildDiagnostics: vscode.DiagnosticCollection) {
+        this._diagnosticsReportedByBuild = buildDiagnostics;
+    }
+
+    public clearDiagnostics() {
+        this._diagnosticsReportedByBuild.clear();
+    }
+
+    public async setBuildDiagnostics(
+        buildDiagnostics: Array<[vscode.Uri, vscode.Diagnostic[]]>,
+        buildOnlyIds: string[]
+    ) {
+        this._allBuildDiagnostics = buildDiagnostics;
+        const displayedBuildDiagnostics = new Array<[vscode.Uri, vscode.Diagnostic[]]>();
+        const allDocuments = vscode.workspace.textDocuments;
+
+        this._allBuildDiagnostics.forEach((fileDiagnostics) => {
+            const uri = fileDiagnostics[0];
+            const diagnosticList = fileDiagnostics[1];
+
+            // Check if the document is open
+            const document = allDocuments.find((d) => this.compareUri(d.uri, uri));
+            const isDocumentOpen = document !== undefined ? !document.isClosed : false;
+
+            // Show the build-only diagnostics
+            displayedBuildDiagnostics.push([
+                uri,
+                BuildDiagnosticsService.filerDiagnosticsFromBuild(diagnosticList, buildOnlyIds, isDocumentOpen),
+            ]);
+        });
+
+        this._diagnosticsReportedByBuild.set(displayedBuildDiagnostics);
+    }
+
+    private compareUri(a: vscode.Uri, b: vscode.Uri): boolean {
+        return a.path.localeCompare(b.path, undefined, { sensitivity: 'accent' }) === 0;
+    }
+
+    public async _onFileOpened(document: vscode.TextDocument, buildOnlyIds: string[]) {
+        const uri = document.uri;
+        const currentFileBuildDiagnostics = this._allBuildDiagnostics?.find(([u]) => this.compareUri(u, uri));
+
+        // The document is now open in the editor and live diagnostics are being shown. Filter diagnostics
+        // reported by the build to show build-only problems.
+        if (currentFileBuildDiagnostics) {
+            const buildDiagnostics = BuildDiagnosticsService.filerDiagnosticsFromBuild(
+                currentFileBuildDiagnostics[1],
+                buildOnlyIds,
+                true
+            );
+            this._diagnosticsReportedByBuild.set(uri, buildDiagnostics);
+        }
+    }
+
+    public static filerDiagnosticsFromBuild(
+        diagnosticList: vscode.Diagnostic[],
+        buildOnlyIds: string[],
+        isDocumentOpen: boolean
+    ): vscode.Diagnostic[] {
+        const configuration = vscode.workspace.getConfiguration();
```

Comment:
you should be able to use the code from https://github.com/dotnet/vscode-csharp/blob/main/src/shared/options.ts to read these instead of directly looking atthe configuration.

Reply:
Is it fine to add to [CommonOptions](https://github.com/dotnet/vscode-csharp/blob/5b856d819588b6719886fb9c6d83538c13aede1a/src/shared/options.ts#L10)? Seems a bit odd, but the best out of the categories available. I could also make a new category if that is more appropriate

Reply:
Ah I didn't realize we didn't have the explicitly there.  I think we should add these to LanguageServerOptions (since O# reads these a little differently).